### ISSUE TITLE
[protobuf] _decode32Bits() bug in ProtobufParser

### ISSUE THREAD

**Issue by marsqing** [FasterXML/jackson-dataformats-binary#85]
```java
protected final int _decode32Bits() throws IOException {
    int ptr = this._inputPtr;
    if(ptr + 3 >= this._inputEnd) {
      return this._slow32();
    } else {
      byte[] b = this._inputBuffer;
      int v = (b[ptr] & 255) + (b[ptr + 1] << 8) + ((b[ptr + 2] & 255) << 16) + (b[ptr + 3] << 24);
      this._inputPtr = ptr + 4;
      return v;
    }
  }
```
bug in the following line
int v = (b[ptr] & 255) + (b[ptr + 1] << 8) + ((b[ptr + 2] & 255) << 16) + (b[ptr + 3] << 24);

If b[prt+1] or b[ptr+3] is negative, << will preserve sign and the overall plus operation will corrupt. Float values like 123.456f will be incorrectly decoded.

_decode64Bits()'s logic is correct.

**Comment by cowtowncoder** [comment]
Thank you again for reporting this. I assume this is with latest version.

**Comment by cowtowncoder** [comment]
Thanks! I added a test (and slightly improved one for `double` too) to catch this, fixed.
Wonder how I didn't notice it; bug is kind of obvious looking at code and I'm familiar with sign extension.
I'll fix this for 2.7, 2.8 and `master` (2.9.0).
